File: android/measure/src/main/java/sh/measure/android/gestures/GestureDetector.kt
package sh.measure.android.gestures

import android.content.Context
import android.view.MotionEvent
import android.view.ViewConfiguration
import sh.measure.android.utils.TimeProvider
import kotlin.math.abs

internal sealed class DetectedGesture {
    internal data class Click(
        val x: Float,
        val y: Float,
        val touchDownTime: Long,
        val touchUpTime: Long,
        val timestamp: Long,
    ) : DetectedGesture()

    internal data class LongClick(
        val x: Float,
        val y: Float,
        val touchDownTime: Long,
        val touchUpTime: Long,
        val timestamp: Long,
    ) : DetectedGesture()

    internal data class Scroll(
        val x: Float,
        val y: Float,
        val endX: Float,
        val endY: Float,
        val direction: Direction,
        val touchDownTime: Long,
        val touchUpTime: Long,
        val timestamp: Long,
    ) : DetectedGesture()
}

internal enum class Direction {
    Down,
    Up,
    Right,
    Left,
}

internal object GestureDetector {
    private var startTouchX = 0f
    private var startTouchY = 0f
    private var startTouchEventTime = 0L
    private var touchSlopPx = 0

    fun detect(
        context: Context,
        motionEvent: MotionEvent,
        timeProvider: TimeProvider,
    ): DetectedGesture? {
        // Ignore multi-touch gestures, this is not supported.
        if (motionEvent.pointerCount > 1) {
            return null
        }
        if (touchSlopPx <= 0) {
            touchSlopPx = ViewConfiguration.get(context).scaledTouchSlop
        }

        if (motionEvent.action == MotionEvent.ACTION_DOWN) {
            startTouchX = motionEvent.x
            startTouchY = motionEvent.y
            startTouchEventTime = motionEvent.eventTime
        }

        if (motionEvent.action == MotionEvent.ACTION_UP) {
            val dt = motionEvent.eventTime - startTouchEventTime
            val dx = abs(startTouchX - motionEvent.x)
            val dy = abs(startTouchY - motionEvent.y)
            return if (dx <= touchSlopPx && dy <= touchSlopPx) {
                if (dt >= ViewConfiguration.getLongPressTimeout()) {
                    DetectedGesture.LongClick(
                        x = motionEvent.x,
                        y = motionEvent.y,
                        touchDownTime = startTouchEventTime,
                        touchUpTime = motionEvent.eventTime,
                        timestamp = timeProvider.now(),
                    )
                } else {
                    DetectedGesture.Click(
                        x = motionEvent.x,
                        y = motionEvent.y,
                        touchDownTime = startTouchEventTime,
                        touchUpTime = motionEvent.eventTime,
                        timestamp = timeProvider.now(),
                    )
                }
            } else {
                DetectedGesture.Scroll(
                    x = startTouchX,
                    y = startTouchY,
                    endX = motionEvent.x,
                    endY = motionEvent.y,
                    touchDownTime = startTouchEventTime,
                    touchUpTime = motionEvent.eventTime,
                    direction = calculateDirection(motionEvent, startTouchX, startTouchY),
                    timestamp = timeProvider.now(),
                )
            }
        }
        return null
    }

    private fun calculateDirection(endEvent: MotionEvent, startX: Float, startY: Float): Direction {
        val diffX = endEvent.x - startX
        val diffY = endEvent.y - startY
        return if (abs(diffX) > abs(diffY)) {
            if (diffX > 0f) {
                Direction.Right
            } else {
                Direction.Left
            }
        } else {
            if (diffY > 0) {
                Direction.Down
            } else {
                Direction.Up
            }
        }
    }
}

File: ios/Sources/MeasureSDK/Swift/Gestures/GestureDetector.swift
//
//  GestureDetector.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 29/09/24.
//

import UIKit

// swiftlint:disable identifier_name
enum DetectedGesture {
    case click(x: CGFloat,
               y: CGFloat,
               touchDownTime: UnsignedNumber,
               touchUpTime: UnsignedNumber,
               target: String?,
               targetId: String?,
               targetFrame: CGRect?)
    case longClick(x: CGFloat,
                   y: CGFloat,
                   touchDownTime: UnsignedNumber,
                   touchUpTime: UnsignedNumber,
                   target: String?,
                   targetId: String?,
                   targetFrame: CGRect?)
    case scroll(x: CGFloat,
                y: CGFloat,
                endX: CGFloat,
                endY: CGFloat,
                direction: Direction,
                touchDownTime: UnsignedNumber,
                touchUpTime: UnsignedNumber,
                target: String?,
                targetId: String?)
}

struct GestureDetector {
    private static var startTouchX: CGFloat = 0
    private static var startTouchY: CGFloat = 0
    private static var startTouchEventTime: UnsignedNumber = 0
    private static var touchSlop: CGFloat = 0
    private static var isScrolling: Bool = false
    private static var target: String?
    private static var targetId: String?
    private static var targetFrame: CGRect?

    static func detect(event: UIEvent, // swiftlint:disable:this cyclomatic_complexity function_body_length
                       in view: UIView,
                       timeProvider: TimeProvider,
                       scaledTouchSlop: CGFloat,
                       longPressTimeout: TimeInterval) -> DetectedGesture? {
        guard let touches = event.allTouches, let touch = touches.first else {
            return nil
        }

        if targetId == nil {
            targetId = touch.view?.accessibilityIdentifier
            targetFrame = touch.view?.frame
        }
        if target == nil, let classForCoder = touch.view?.classForCoder {
            target = NSStringFromClass(classForCoder)
        }

        if touchSlop == 0 {
            touchSlop = scaledTouchSlop
        }

        let location = touch.location(in: view)

        switch touch.phase {
        case .began:
            startTouchX = location.x
            startTouchY = location.y
            startTouchEventTime = UnsignedNumber(timeProvider.millisTime)
            isScrolling = false
        case .moved:
            // Movement detected, check if it is a scroll gesture
            let dx = abs(startTouchX - location.x)
            let dy = abs(startTouchY - location.y)

            if (dx > touchSlop || dy > touchSlop) && !isScrolling {
                // Once the movement exceeds the touchSlop, consider it a scroll
                isScrolling = true
            }
        case .ended:
            if !isScrolling {
                let currentTime = UInt64(timeProvider.millisTime)
                if currentTime > startTouchEventTime {
                    let dt = currentTime - startTouchEventTime
                    let dx = abs(startTouchX - location.x)
                    let dy = abs(startTouchY - location.y)
                    if dx <= touchSlop && dy <= touchSlop {
                        if TimeInterval(dt) >= longPressTimeout {
                            let longClick: DetectedGesture = .longClick(
                                x: location.x,
                                y: location.y,
                                touchDownTime: startTouchEventTime,
                                touchUpTime: UnsignedNumber(timeProvider.millisTime),
                                target: target,
                                targetId: targetId,
                                targetFrame: targetFrame)
                            target = nil
                            targetId = nil
                            return longClick
                        } else {
                            let click: DetectedGesture = .click(
                                x: location.x,
                                y: location.y,
                                touchDownTime: startTouchEventTime,
                                touchUpTime: UnsignedNumber(timeProvider.millisTime),
                                target: target,
                                targetId: targetId,
                                targetFrame: targetFrame)
                            target = nil
                            targetId = nil
                            return click
                        }
                    }
                }
            } else {
                let scroll: DetectedGesture = .scroll(
                    x: startTouchX,
                    y: startTouchY,
                    endX: location.x,
                    endY: location.y,
                    direction: calculateScrollDirection(endX: location.x, endY: location.y, startX: startTouchX, startY: startTouchY),
                    touchDownTime: startTouchEventTime,
                    touchUpTime: UnsignedNumber(timeProvider.millisTime),
                    target: target,
                    targetId: targetId)
                target = nil
                targetId = nil
                return scroll
            }
        default:
            return nil
        }

        return nil
    }

    private static func calculateScrollDirection(endX: CGFloat, endY: CGFloat, startX: CGFloat, startY: CGFloat) -> Direction {
        let diffX = endX - startX
        let diffY = endY - startY
        if abs(diffX) > abs(diffY) {
            return diffX > 0 ? .right : .left
        } else {
            return diffY > 0 ? .down : .up
        }
    }
}
// swiftlint:enable identifier_name
